winlogbeat/module/pipeline.go

```diff
@@ -0,0 +1,211 @@
+		return err
+	}
+
+	for _, pipeline := range pipelines {
+		if err := fileset.AdaptPipelineForCompatibility(version, pipeline.id, pipeline.contents, log); err != nil {
+			return fmt.Errorf("failed to adapt pipeline with backwards compatibility changes for version %s: %w", version.String(), err)
+		}
+		f, err := os.Create(filepath.Join(directory, pipeline.id+".json"))
+		if err != nil {
+			return fmt.Errorf("unable to create file to export pipeline to: %+v", err)
+		}
+		enc := json.NewEncoder(f)
+		enc.SetEscapeHTML(false)
+		err = enc.Encode(pipeline.contents)
+		if err != nil {
+			return fmt.Errorf("unable to JSON encode pipeline %s: %+v", f.Name(), err)
+		}
+	}
+	return nil
+}
+
+// readAll reads pipelines from the the embedded filesystem and
+// returns a slice of pipelines suitable for sending to Elasticsearch
+// with load.
+func readAll(info beat.Info) (pipelines []pipeline, err error) {
+	p, err := readDir(".", info)
+	if err == errNoFS {
+		return nil, nil
+	}
+	return p, err
+}
+
+func readDir(dir string, info beat.Info) (pipelines []pipeline, err error) {
+	if PipelinesFS == nil {
+		return nil, errNoFS
+	}
+	dirEntries, err := PipelinesFS.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	for _, de := range dirEntries {
+		if de.IsDir() {
+			subPipelines, err := readDir(path.Join(dir, de.Name()), info)
+			if err != nil {
+				return nil, err
+			}
+			pipelines = append(pipelines, subPipelines...)
+			continue
+		}
+		p, err := readFile(path.Join(dir, de.Name()), info)
+		if err == errNoFS {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		pipelines = append(pipelines, p)
+	}
+	return pipelines, nil
+}
+
+func readFile(filename string, info beat.Info) (p pipeline, err error) {
+	if PipelinesFS == nil {
+		return pipeline{}, errNoFS
+	}
+	contents, err := PipelinesFS.ReadFile(filename)
+	if err != nil {
+		return pipeline{}, err
+	}
+	updatedContent, err := applyTemplates(info.IndexPrefix, info.Version, filename, contents)
+	if err != nil {
+		return pipeline{}, err
+	}
+	p = pipeline{
+		id:       fileset.FormatPipelineID(info.IndexPrefix, "default", "default", filename, info.Version),
+		contents: updatedContent,
+	}
+	return p, nil
+}
+
+// Load uses esClient to load pipelines to Elasticsearch
+// cluster.  Will only overwrite existing pipelines if
+// overwritePipelines is true.
+func load(esClient *eslegclient.Connection, pipelines []pipeline, overwritePipelines bool, log *logp.Logger) (err error) {
+	var pipelineIDsLoaded []string
+
+	for _, pipeline := range pipelines {
+		err = fileset.LoadPipeline(esClient, pipeline.id, pipeline.contents, overwritePipelines, log)
+		if err != nil {
+			err = fmt.Errorf("error loading pipeline %s: %v", pipeline.id, err)
+			break
+		}
+		pipelineIDsLoaded = append(pipelineIDsLoaded, pipeline.id)
+	}
+
+	if err != nil {
+		errs := multierror.Errors{err}
+		for _, pipelineID := range pipelineIDsLoaded {
+			err = fileset.DeletePipeline(esClient, pipelineID)
```

Comment:
This cleanup-on-error behavior is a little surprising to me. I recommend adding this to the godoc.

Reply:
I'll add to the godoc.  This behavior is copied from filebeat to mimic its behavior
https://github.com/elastic/beats/blob/e6e65aa92fe355c95789691ebf5a3bcecaf5b4ea/filebeat/fileset/pipelines.go#L103